crates/telio-pmtu/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+#[cfg(any(target_os = "linux", target_os = "android"))]
+#[path = "linux.rs"]
+mod platform;
+
+use std::{net::IpAddr, ops::Range, sync::Arc, time::Duration};
+
+use tokio::task::JoinHandle;
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+pub use platform::PMTUSocket;
+
+// We assume that 1000 is always supported.
+//1500 is the Ethernet MTU so we're not expecting PMTU to be higher than that.
+#[allow(unused)]
+const PMTU_RANGE: Range<u32> = 1000..1501;
+
+#[allow(unused)]
+const ICMP_ECHO_REQ_IDENT: u16 = 0x6C74; // "lt" (libtelio) in hex
+
+#[allow(unused)]
+const INTERVAL: Duration = Duration::from_secs(10 * 60); // 10 mins
+
+pub struct Entity {
+    #[allow(unused)]
+    socket_pool: Arc<telio_sockets::SocketPool>,
+    task: Option<JoinHandle<()>>,
+}
+
+impl Entity {
+    pub fn new(socket_pool: Arc<telio_sockets::SocketPool>) -> Self {
+        Self {
+            socket_pool,
+            task: None,
+        }
+    }
+
+    pub async fn run(&mut self, host: IpAddr) {
+        self.stop().await;
+        self.start(host);
+    }
+
+    pub async fn stop(&mut self) {
+        if let Some(task) = self.task.take() {
+            telio_utils::telio_log_debug!("Stopping PMTU discovery task");
+            task.abort();
+            let _ = task.await;
+        }
+    }
+
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    fn start(&mut self, host: IpAddr) {
+        telio_utils::telio_log_debug!("Starting PMTU discovery task");
+
+        use std::os::fd::AsRawFd;
+
+        use telio_lana::*;
+
+        async fn probe(
+            socket_pool: &telio_sockets::SocketPool,
+            host: IpAddr,
+        ) -> std::io::Result<u32> {
+            let sock = PMTUSocket::new(host)?;
+            telio_utils::telio_log_debug!("Making PMTU socket external");
+            socket_pool.make_external(sock.as_raw_fd());
+            sock.probe_pmtu().await
+        }
+
+        let pool = self.socket_pool.clone();
+        let task = tokio::spawn(async move {
+            let mut timer = tokio::time::interval(INTERVAL);
+
+            loop {
+                timer.tick().await;
```

Comment:
Could you remind me what is `tokio::time::interval(...)` behavior in terms of sleep? E.g. if the process/device is sleeping, will "tick()" drop missed events, or will it try to "catch-up"?

Reply:
It will catch up, although it's configurable. Sleep just speels and `tick()` does more or less `sleep(INTERVAL - <time it takes for loop to turn once>)`

Reply:
:thinking: I think it would be nice to configure it to skip the events, as otherwise we will be spamming many PMTU detection events after waking up from sleep.

Reply:
Done